vsphere/check.py

```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
+
+                    vms = []
+
+                    for vm in c.vm:
+                        vms.append(vm.name)
+
+                    topology_tags["vms"] = vms
+
+                    hostname = None
+                elif isinstance(c, vim.Datacenter):
+                    datastores = []
+                    for datastore in c.datastore:
+                        datastores.append(datastore.name)   # datastore._moId - GUID of store
+                    topology_tags["topo_type"] = VSPHERE_COMPONENT_TYPE.DATACENTER
+                    topology_tags["datastores"] = datastores
+                    topology_tags["name"] = c.name
+                    topology_tags["id"] = c._moId
+                    topology_tags["layer"] = TOPOLOGY_LAYERS.DATACENTER
+                    topology_tags["domain"] = domain
+
+                    computeresources = []
+                    clustercomputeresources = []
+
+                    for computeres in c.hostFolder.childEntity:
+                        if isinstance(computeres, vim.ComputeResource):
+                            computeresources.append(computeres.name)
+                        elif isinstance(computeres, vim.CloudComputeResource):
+                            clustercomputeresources.append(computeres.name)
+
+                    topology_tags["computeresources"] = computeresources
+                    topology_tags["clustercomputeresources"] = clustercomputeresources
+
+                    add_label_pair(labels, "name", topology_tags["name"])
+
+                    hostname = None
+                topology_tags["labels"] = labels
+                obj_list.append(dict(mor_type=vimtype, mor=c, hostname=hostname, topo_tags = topology_tags))
+
+        return obj_list
+
+    def get_topologyitems_sync(self, instance, tags=[], regexes=None, include_only_marked=False):
+        server_instance = self._get_server_instance(instance)
+        content = server_instance.RetrieveContent()
+        domain = instance["host"]  # candidate also name
+
+        vms = self._vsphere_objs(content, "vm", domain)
+        hosts = self._vsphere_objs(content, "host", domain)
+        datacenters = self._vsphere_objs(content, "datacenter", domain)
+        datastores = self._vsphere_objs(content, "datastore", domain)
+        clustercomputeresource = self._vsphere_objs(content, "clustercomputeresource", domain)
+        computeresource = self._vsphere_objs(content, "computeresource", domain)
+
+        return {
+            "vms": vms,
+            "hosts": hosts,
+            "datacenters": datacenters,
+            "datastores": datastores,
+            "clustercomputeresource": clustercomputeresource,
+            "computeresource": computeresource
+        }
+
+    def collect_topology(self,instance):
+
+        def build_id(vsphere_url, object_type, object_name):
+            return "urn://vsphere/{0}/{1}/{2}".format(vsphere_url, object_type, object_name)
```

Comment:
Can you suggest proposed ? I am following ideas from azure topology here

Reply:
@whazenberg the URN used is inline with what we have for AWS / Azure for example, however I would propose that we format it as: `urn: vsphere:/{instance_id/url}/{object_type}/{object_name}` It's really important that the URN uniquely describes the component across all instances, so sometimes it can be a lot of info, however if you take a look at the Azure / AWS equivalents, it also uses a similar URI based URN. For example one URN we have in Azure is: `urn:azure:/SUBSCRIPTIONS/F385376B-50BF-42D5-B0F4-C97CF497F82F/RESOURCEGROUPS/STSSTACKPACKTEST-RG/PROVIDERS/MICROSOFT.WEB/SITES/STSAGENT/FUNCTIONS/EVENTHUBCHANGESFUNCTION`

Reply:
Ack, let's change the URN format to what Zandré proposed?

Reply:
i.e. add `:` after vsphere?

Zandre says:

```
however I would propose that we format it as: urn:vsphere:/{instance_id/url}/{object_type}/{object_name}
```

we have: 

```
return "urn://vsphere/{0}/{1}/{2}".format(vsphere_url, object_type, object_name)

```

difference is `:`

Reply:
@whazenberg URN is fixed now.